internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,146 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package speakerlist
+
+import (
+	"crypto/sha256"
+	golog "log"
+	"strconv"
+	"time"
+
+	"go.universe.tf/metallb/internal/k8s"
+
+	gokitlog "github.com/go-kit/kit/log"
+	"github.com/hashicorp/memberlist"
+)
+
+// SpeakerList gives you the list of healthy speaker
+type SpeakerList struct {
+	l         gokitlog.Logger
+	client    *k8s.Client
+	stopCh    chan struct{}
+	namespace string
+	labels    string
+	// following fields are nil when MemberList is disabled
+	mlEventCh chan memberlist.NodeEvent
+	mList     *memberlist.Memberlist
+}
+
+// New creates a new SpeakerList
+func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace, labels string, stopCh chan struct{}) (*SpeakerList, error) {
+	sl := SpeakerList{
+		l:         logger,
+		stopCh:    stopCh,
+		namespace: namespace,
+		labels:    labels,
+	}
+
+	if namespace == "" || labels == "" || bindAddr == "" {
+		logger.Log("op", "startup", "msg", "Not starting fast dead node detection (MemberList), need ml-bindaddr / ml-labels / ml-namespace config")
+		return &sl, nil
+	}
+
+	mconfig := memberlist.DefaultLANConfig()
+	// mconfig.Name MUST be spec.nodeName, as we will match it against Enpoints nodeName in usableNodes()
+	mconfig.Name = nodeName
+	mconfig.BindAddr = bindAddr
+	if bindPort != "" {
+		mlport, err := strconv.Atoi(bindPort)
+		if err != nil {
+			sl.l.Log("op", "startup", "error", "unable to parse ml-bindport", "msg", err)
+			return nil, err
+		}
+		mconfig.BindPort = mlport
+		mconfig.AdvertisePort = mlport
+	}
+	loggerout := gokitlog.NewStdlibAdapter(gokitlog.With(sl.l, "component", "MemberList"))
+	mconfig.Logger = golog.New(loggerout, "", golog.Lshortfile)
+	if secret == "" {
+		sl.l.Log("op", "startup", "warning", "no ml-secret-key set, MemberList traffic will not be encrypted")
+	} else {
+		sha := sha256.New()
+		mconfig.SecretKey = sha.Sum([]byte(secret))[:16]
+	}
+	// ChannelEventDelegate hint that it should not block, so make mlEventCh 'big'
```

Comment:
From https://pkg.go.dev/github.com/hashicorp/memberlist@v0.1.7?tab=doc#ChannelEventDelegate:

> Care must be taken that events are processed in a timely manner from the channel, since this delegate will block until an event can be sent. 

I'm not 100% sure what this would cause, maybe the node might be considered down. But:
 * This is the code we already have in master, we are just moving it in the commit
 * It is probably not a big issue, because events are processed quite fast: https://github.com/metallb/metallb/pull/662/files#diff-b7338554f9f6fe870766bf340b113a65R291-R300

I can create an issue to see the pros/cons of moving to https://pkg.go.dev/github.com/hashicorp/memberlist@v0.1.7?tab=doc#EventDelegate

Maybe @champtar is already aware of the pros/cons?

Reply:
From the top of my head, I think I've rework how I use those events to make sure I don't block, but it was better (or easier) to use the channel that the function call.

Reply:
@rata I'm OK with leaving this for now since we're just moving this code around, however I'd rather not have bits in our codebase we don't understand, and in general this feels a bit hacky to me. I suggest leaving a TODO with a link to an issue so that we don't forget to handle this later.

Reply:
Added a TODO comment pointing to this issue: https://github.com/metallb/metallb/issues/716